x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Chunk.java

```diff
@@ -159,78 +204,101 @@ public Expression replaceChildren(List<Expression> newChildren) {
+    @Evaluator(extraName = "BytesRef")
+    static void process(BytesRefBlock.Builder builder, BytesRef str, int numChunks, int chunkSize) {
+        var settings = new SentenceBoundaryChunkingSettings(chunkSize, 0);
+        var chunks = chunkText(str.utf8ToString(), settings);
+        if (chunks.size() > numChunks) {
+            chunks = chunks.subList(0, numChunks);
         }
+        appendBytesRefs(builder, chunks);
     }
 
-    public static List<String> chunkText(String content, ChunkingSettings chunkingSettings, int numChunks) {
+    public static List<String> chunkText(String content, ChunkingSettings chunkingSettings) {
         Chunker chunker = ChunkerBuilder.fromChunkingStrategy(chunkingSettings.getChunkingStrategy());
-        return chunker.chunk(content, chunkingSettings)
-            .stream()
-            .map(offset -> content.substring(offset.start(), offset.end()))
-            .limit(numChunks > 0 ? numChunks : DEFAULT_NUM_CHUNKS)
-            .toList();
+        return chunker.chunk(content, chunkingSettings).stream().map(offset -> content.substring(offset.start(), offset.end())).toList();
+    }
+
+    private static void appendBytesRefs(BytesRefBlock.Builder builder, List<String> texts) {
+        boolean multivalued = texts.size() > 1;
+        if (multivalued) builder.beginPositionEntry();
+        texts.forEach(t -> builder.appendBytesRef(new BytesRef(t.trim())));
+        if (multivalued) builder.endPositionEntry();
     }
 
     @Override
     public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) return false;
         Chunk chunk = (Chunk) o;
-        return Objects.equals(field(), chunk.field()) && Objects.equals(options(), chunk.options());
+        return Objects.equals(field(), chunk.field())
+            && Objects.equals(query(), chunk.query())
+            && Objects.equals(options(), chunk.options());
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(field(), options());
+        return Objects.hash(field(), query(), options());
     }
 
     @Override
     public EvalOperator.ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
 
         Map<String, Object> optionsMap = new HashMap<>();
         if (options() != null) {
-            Options.populateMap(((MapExpression) options), optionsMap, source(), SECOND, ALLOWED_OPTIONS);
+            Options.populateMap(((MapExpression) options()), optionsMap, source(), SECOND, ALLOWED_OPTIONS);
         }
 
         int numChunks = (Integer) optionsMap.getOrDefault(NUM_CHUNKS, DEFAULT_NUM_CHUNKS);
         int chunkSize = (Integer) optionsMap.getOrDefault(CHUNK_SIZE, DEFAULT_CHUNK_SIZE);
 
+        if (query() != null) {
+            return new ChunkBytesRefRescoreEvaluator.Factory(
+                source(),
+                toEvaluator.apply(field()),
+                toEvaluator.apply(query()),
+                toEvaluator.apply(new Literal(source(), numChunks, DataType.INTEGER)),
```

Comment:
Can we avoid doing the evaluation, as discussed previously? We should be able to retrieve directly the value instead of having to evaluate a Literal with the value.